File: src-tauri/src/map_maker/chunk_64.rs
// ---- Landscape chunk u32 bit ordering
//
// Bits | Name
// -----------------
// 2     | rotation
// 1     | x-mirror
// 1     | z-mirror
// 10    | x-position
// 10    | z-position
// 10    | y-position
// 8     | set ID
// 10    | mesh ID
//
//   0                                          1
//   |--------------------------------------|   |-----------------------------------------------|
//    31             19           9              31         23           13           3   2   1
//    000000000000   0000000000   0000000000 <-> 00000000   0000000000   0000000000   0   0   00
//   |------------| |----------| |----------|   |--------| |----------| |----------| |-| |-| |--|
//   |              |            |              |          |            |            |   |   ┕ rotation
//   |              |            |              |          |            |            |   ┕ x-mirror
//   |              |            |              |          |            |            ┕ z-mirror
//   |              |            |              |          |            ┕ x-position
//   |              |            |              |          ┕ z-position
//   |              |            |              ┕ set ID
//   |              |            ┕ y-position
//   |              ┕ mesh ID
//   ┕ NOT USED

pub struct LandscapeChunk64 {
    pub id: u32,
    pub mesh_id: u32,
    pub set_id: u32,
    pub y: u32,
    pub z: u32,
    pub x: u32,
    pub mirror: (u32, u32),
    pub rotation: u32,
}

impl LandscapeChunk64 {
    pub fn new(chunk: (u32, u32), map_width: u32, map_depth: u32) -> LandscapeChunk64 {
        LandscapeChunk64 {
            id: get_id(chunk, map_width, map_depth),
            mesh_id: get_mesh_id(chunk),
            set_id: get_set_id(chunk),
            y: get_y_pos(chunk),
            z: get_z_pos(chunk),
            x: get_x_pos(chunk),
            mirror: (get_x_mirror(chunk), get_z_mirror(chunk)),
            rotation: get_rotation(chunk),
        }
    }

    pub fn pack(self) -> (u32, u32) {
        create(
            self.mesh_id,
            self.set_id,
            self.x,
            self.z,
            self.y,
            self.mirror,
            self.rotation,
        )
    }
}

pub fn create(
    mesh_id: u32,
    set_id: u32,
    x_pos: u32,
    z_pos: u32,
    y_pos: u32,
    mirror: (u32, u32),
    rotation: u32,
) -> (u32, u32) {
    let mut chunk: (u32, u32) = (0, 0);
    set_mesh_id(&mut chunk, mesh_id);
    set_set_id(&mut chunk, set_id);
    set_x_pos(&mut chunk, x_pos);
    set_z_pos(&mut chunk, z_pos);
    set_y_pos(&mut chunk, y_pos);
    set_x_mirror(&mut chunk, mirror.0);
    set_z_mirror(&mut chunk, mirror.1);
    set_rotation(&mut chunk, rotation);
    chunk
}

pub fn get_id(chunk: (u32, u32), map_width: u32, map_depth: u32) -> u32 {
    let x = get_x_pos(chunk);
    let y = get_y_pos(chunk);
    let z = get_z_pos(chunk);
    x + z * map_width + y * map_width * map_depth
}

pub fn set_mesh_id(chunk: &mut (u32, u32), mesh_id: u32) {
    chunk.0 = (chunk.0 & 0xFFF003FF) | ((mesh_id & 0x3FF) << 10);
}
pub fn get_mesh_id(chunk: (u32, u32)) -> u32 {
    (chunk.0 & 0xFFC00) >> 10
}

pub fn set_set_id(chunk: &mut (u32, u32), set_id: u32) {
    chunk.1 = (chunk.1 & 0xFFFFFF) | ((set_id & 0xFF) << 24);
}
pub fn get_set_id(chunk: (u32, u32)) -> u32 {
    (chunk.1 & 0xFF000000) >> 24
}

pub fn set_y_pos(chunk: &mut (u32, u32), y_pos: u32) {
    chunk.0 = (chunk.0 & 0xFFFFFC00) | (y_pos & 0x3FF);
}
pub fn get_y_pos(chunk: (u32, u32)) -> u32 {
    chunk.0 & 0x3FF
}

pub fn set_z_pos(chunk: &mut (u32, u32), z_pos: u32) {
    chunk.1 = (chunk.1 & 0xFF003FFF) | ((z_pos & 0x3FF) << 14);
}
pub fn get_z_pos(chunk: (u32, u32)) -> u32 {
    (chunk.1 & 0xFFC000) >> 14
}

pub fn set_x_pos(chunk: &mut (u32, u32), x_pos: u32) {
    chunk.1 = (chunk.1 & 0xFFFFFC0F) | ((x_pos & 0x3FF) << 4);
}
pub fn get_x_pos(chunk: (u32, u32)) -> u32 {
    (chunk.1 & 0x7F0) >> 4
}

pub fn set_z_mirror(chunk: &mut (u32, u32), z_mirror: u32) {
    chunk.1 = (chunk.1 & 0xFFFFFFF7) | ((z_mirror & 0x1) << 3);
}
pub fn get_z_mirror(chunk: (u32, u32)) -> u32 {
    (chunk.1 & 0x8) >> 3
}

pub fn set_x_mirror(chunk: &mut (u32, u32), x_mirror: u32) {
    chunk.1 = (chunk.1 & 0xFFFFFFFB) | ((x_mirror & 0x1) << 2);
}
pub fn get_x_mirror(chunk: (u32, u32)) -> u32 {
    (chunk.1 & 0x4) >> 2
}

pub fn set_rotation(chunk: &mut (u32, u32), rotation: u32) {
    chunk.1 = (chunk.1 & 0xFFFFFFFC) | (rotation & 0b11);
}
pub fn get_rotation(chunk: (u32, u32)) -> u32 {
    chunk.1 & 0x3
}

File: src/map-maker/chunk-64.ts
export type Chunk64Mirror = [0 | 1, 0 | 1];

export class Chunk64 {
    id: number;
    meshId: number;
    setId: number;
    y: number;
    z: number;
    x: number;
    mirror: Chunk64Mirror;
    rotation: number;

    constructor(
        chunk: [number, number],
        public mapWidth: number,
        public mapDepth: number,
    ) {
        this.id = ChunkManipulation64.getId(chunk, mapWidth, mapDepth);
        this.meshId = ChunkManipulation64.getMeshId(chunk);
        this.setId = ChunkManipulation64.getSetId(chunk);
        this.y = ChunkManipulation64.getYPos(chunk);
        this.z = ChunkManipulation64.getZPos(chunk);
        this.x = ChunkManipulation64.getXPos(chunk);
        this.mirror = [
            ChunkManipulation64.getXMirror(chunk),
            ChunkManipulation64.getZMirror(chunk),
        ];
        this.rotation = ChunkManipulation64.getRotation(chunk);
    }

    pack(): [number, number] {
        return ChunkManipulation64.create(
            this.meshId,
            this.setId,
            this.x,
            this.z,
            this.y,
            this.mirror,
            this.rotation,
        );
    }
}

export class ChunkManipulation64 {
    static create(
        mesh_id: number,
        set_id: number,
        x_pos: number,
        z_pos: number,
        y_pos: number,
        mirror: [number, number],
        rotation: number,
    ): [number, number] {
        const chunk: [number, number] = [0, 0];
        this.setMeshId(chunk, mesh_id);
        this.setSetId(chunk, set_id);
        this.setXPos(chunk, x_pos);
        this.setZPos(chunk, z_pos);
        this.setYPos(chunk, y_pos);
        this.setXMirror(chunk, mirror[0]);
        this.setZMirror(chunk, mirror[1]);
        this.setRotation(chunk, rotation);
        return chunk;
    }

    static getId(
        chunk: [number, number],
        map_width: number,
        map_depth: number,
    ): number {
        const x = this.getXPos(chunk);
        const y = this.getYPos(chunk);
        const z = this.getZPos(chunk);
        return x + z * map_width + y * map_width * map_depth;
    }

    static setMeshId(chunk: [number, number], mesh_id: number) {
        chunk[0] = (chunk[0] & 0xfff003ff) | ((mesh_id & 0x3ff) << 10);
    }
    static getMeshId(chunk: [number, number]): number {
        return (chunk[0] & 0xffc00) >> 10;
    }

    static setSetId(chunk: [number, number], set_id: number) {
        chunk[1] = (chunk[1] & 0xffffff) | ((set_id & 0xff) << 24);
    }
    static getSetId(chunk: [number, number]): number {
        return (chunk[1] & 0xff000000) >> 24;
    }

    static setYPos(chunk: [number, number], y_pos: number) {
        chunk[0] = (chunk[0] & 0xfffffc00) | (y_pos & 0x3ff);
    }
    static getYPos(chunk: [number, number]): number {
        return chunk[0] & 0x3ff;
    }

    static setZPos(chunk: [number, number], z_pos: number) {
        chunk[1] = (chunk[1] & 0xff003fff) | ((z_pos & 0x3ff) << 14);
    }
    static getZPos(chunk: [number, number]): number {
        return (chunk[1] & 0xffc000) >> 14;
    }

    static setXPos(chunk: [number, number], x_pos: number) {
        chunk[1] = (chunk[1] & 0xfffffc0f) | ((x_pos & 0x3ff) << 4);
    }
    static getXPos(chunk: [number, number]): number {
        return (chunk[1] & 0x7f0) >> 4;
    }

    static setZMirror(chunk: [number, number], z_mirror: 0 | 1) {
        chunk[1] = (chunk[1] & 0xfffffff7) | ((z_mirror & 0x1) << 3);
    }
    static getZMirror(chunk: [number, number]): number {
        return (chunk[1] & 0x8) >> 3;
    }

    static setXMirror(chunk: [number, number], x_mirror: 0 | 1) {
        chunk[1] = (chunk[1] & 0xfffffffb) | ((x_mirror & 0x1) << 2);
    }
    static getXMirror(chunk: [number, number]): number {
        return (chunk[1] & 0x4) >> 2;
    }

    static setRotation(chunk: [number, number], rotation: number) {
        chunk[1] = (chunk[1] & 0xfffffffc) | (rotation & 0b11);
    }
    static getRotation(chunk: [number, number]): number {
        return chunk[1] & 0x3;
    }
}
